Add a replicate_for option to the ILM searchable_snapshot action

joegallo opened the pull request:
Adds a `replicate_for` option (expressed as a time value) to the ILM `searchable_snapshot` action. If that option is present, then the searchable snapshot will be mounted with one replica (rather than zero replicas), and ILM execution will wait in the searchable snapshot action until the `replicate_for` time value has passed. Once the time value has passed, then the replica will be removed and execution will continue.

As an example, this makes it possible to have a `frozen` phase that lasts for 60 days, but to have 14 days worth of frozen data mounted with replicas.

joegallo committed:
Add a transport version

joegallo committed:
Add replicate_for to the SearchableSnapshotAction

joegallo committed:
Add a step to wait for the replicate_for time to pass

joegallo committed:
Wire waiting step into searchable_snapshot action

joegallo committed:
Add whole-policy replicate_for validation

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -473,6 +474,77 @@ static void validateDownsamplingIntervals(Collection<Phase> phases) {
         }
     }
 
+    static void validateReplicateFor(Collection<Phase> phases) {
```
Can you add a javadoc snippet for what this validation is going to this method?

joegallo replied:
Sure thing, a9e600154e84fdcc727aa3e888dfaa399f748670.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitUntilReplicateForTimePassesStep.java:
```diff
@@ -0,0 +1,104 @@
+                false,
+                new SingleMessageFieldInfo(
+                    Strings.format(
+                        "Waiting until the replicate_for time [%s] has elapsed for index [%s] before removing replicas.",
```
Think it'd be worth also including how long the calculated time is? So that a user can see "oh, it's waiting for 7 days and it's at 6 days right now"?

joegallo replied:
I didn't want to have the message change over time because in the degenerate case that would mean that each ILM run for one of these indices would also be a cluster state update. If we limit the granularity of the message in a meaningful way, though, then it would continue to be lazy -- I'll see what I can do.

joegallo replied:
```
      "step_info" : {
        "message" : "Waiting until the replicate_for time [2d] has elapsed for index [restored-my-index-1] before removing replicas. Approximately [1d] remaining."
      },
```

Using `remaining.toHumanReadableString(0)` gives us the above string, but the behavior is a floor on the time unit -- so in this case there's something like 1.99 days remaining, but it's saying `1d`. I like some aspects of this solution -- we'll push through one cluster state change to update the `step_info` `message` String each day, then for about one day we'd do updates once an hour, and then finally for one hour we'd do an update on every ILM run.

A cleverer solution would do rounding rather than flooring, and it would fix the minimum granularity to hours -- maybe the minimum answer should be (1h, 0.5h, 0.25h hours)... or maybe it should round up and down on just an hour (so at the end it's ~= `1h` and then ~= `0h`, which now that I'm typing it I would actually say I'm pretty taken with).

joegallo replied:
What do you think about the somewhat-clever solution in 4a46546f81656c1f335bc3f9fedc4551c007f685, @dakrone?

dakrone reviewed (commented):
Provisionally looks good to me!

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

joegallo committed:
Add a javadoc for validateReplicateFor

joegallo committed:
Add a mechanism for approximating the time remaining

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

joegallo committed:
Fix a typo

joegallo committed:
Handle a null edge case

It shouldn't ever happen, but ever is a long time.

joegallo committed:
Add a small explanatory comment

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

joegallo committed:
Fix the validation logic to handle more scenarios

and add tests of the validation logic

joegallo committed:
Change the behavior of approximateTimeRemaining

and update the tests. Also, rewrite the tests to correctly delineate
actual and expected values (doh!).

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

joegallo committed:
Add an integration test

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

joegallo committed:
Update docs/changelog/119003.yaml

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

joegallo commented:
I've marked this as 'Ready for Review' and I've removed the 'WIP' label. I still need to write the documentation for this -- and I will do so on this PR before we merge it. Feel free to review the code and add comments where necessary, but I encourage you to withhold a ✅ until the docs have been written (which I will do tomorrow).

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -473,6 +474,109 @@ static void validateDownsamplingIntervals(Collection<Phase> phases) {
+                    if (phaseMinAge != TimeValue.ZERO && minAgeDeltaMillis < firstReplicateFor.millis()) {
+                        throw new IllegalArgumentException(
+                            Strings.format(
+                                "the value for replicate_for [%s] must be less than or equal to the difference in min_age for "
+                                    + "the current phase and subsequent phases, but the [%s] phase has a min_age of [%s] "
+                                    + "and the [%s] phase has a min_age of [%s], so the difference was only [%s]",
```
I think we can make this clearer to people that aren't familiar with the technical wording, perhaps something like:

"The time a searchable snapshot is replicated in replicate_for […] may not exceed the time the next phase is configured to begin. Based on the min_age of […] for the […] phase, the maximum time the snapshot could be replicated is […]."

What do you think?

joegallo replied:
I went with your message (approximately!) in 39e6660644a, but I reworded it very slightly -- feel free to insist on different wording than what I happened to go with.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -60,13 +65,19 @@ public class SearchableSnapshotAction implements LifecycleAction {
         PARSER.declareInt(ConstructingObjectParser.optionalConstructorArg(), TOTAL_SHARDS_PER_NODE);
+        PARSER.declareField(
+            ConstructingObjectParser.optionalConstructorArg(),
+            p -> TimeValue.parseTimeValue(p.textOrNull(), REPLICATE_FOR.getPreferredName()),
```
We should validate that this is neither `TimeValue.ZERO` nor `TimeValue.MINUS_ONE`, to avoid someone passing `replicate_for: -1` in the config

joegallo replied:
Sure thing, I did that in 8a10b7f9e3c.

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/WaitUntilReplicateForTimePassesStepTests.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    public void testEvaluateCondition() {
+        // a mutable box that we can put Instants into
+        final Instant[] returnVal = new Instant[1];
```
Can we use `AtomicReference` for this instead? This seems like a wallhack to me, and is less readable :)

joegallo replied:
Absolutely, fixed in d88150a8fdd.

dakrone reviewed (commented):
Looks good to me in general, I left a few comments, but nothing major.

joegallo committed:
Add validation of the replicate_for time

It can be null, but if it isn't, then it need to be a positive amount
of time (not -1 or 0).

joegallo committed:
Drop this unused logger

joegallo committed:
Use an AtomicReference for the mutable box

joegallo committed:
Fix a test bug

joegallo committed:
Reword this validation message string

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

dakrone reviewed (approved):
LGTM, are you going to do docs in this PR, or a subsequent one?

joegallo commented:
> LGTM, are you going to do docs in this PR, or a subsequent one?

I'll do them on this PR first thing tomorrow -- thanks for the code review, and I'll ping you for one last look at the docs before I merge this.

joegallo committed:
Merge branch 'main' into ilm-searchable-snapshot-replicate-for

joegallo committed:
Add docs for replicate_for

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

joegallo committed:
Delete docs/changelog/119003.yaml

joegallo commented:
https://elasticsearch_bk_119003.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/ilm-searchable-snapshot.html

@dakrone what do you think of this for the docs?

joegallo committed:
Update docs/changelog/119003.yaml

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

dakrone reviewed (approved):
Docs LGTM also

joegallo commented:
CI passed but github missed the update notification, I'm merging this.

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/120956)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/120956">](https://github.com/elastic/elasticsearch/pull/120956) |
